### PR TITLE
[WIP] Tempered alt

### PR DESCRIPTION
**Pull request by JesusTorrado**

### CODE REVIEW

**Comment by JesusTorrado** [comment]
Another new thing in the branch:

Now can automatically do the MPI gathering, skipping, concatenating and (optionally) GetDist conversion at once:

    mcmc.products(
        combined=True,  # returns concatenated MPI chains
        skip_samples=0.3,  # skips initial fraction (applied prev to concatenation)
        to_getdist=True,  # returns MCSamples (safely combined with the other two args)
    )

Thoughts? If you find it useful, I would update the docs to use this by default (simpler). In particular, we should promote using the new `Collection.to_getdist()` instead of `MCSamplesFromCobaya`, because it's safer.

**Comment by cmbant** [comment]
Sounds fine to me, though would be a potentially breaking change having getdist by default. I might be inclined to make a separate method, e.g. mcmc.getdist_chains(...) or similar returning the getdist object directly, avoiding to have to "sample" index the dict. (singular "sample" has always been a bit odd)

Btw, I don't think we have any function to get mcsamples with the multi-chain information (e.g. for running Getdist's other convergence tests).

**Comment by cmbant** [comment]
Currently the updated getdist does not run with master-branch cobaya, e.g. test_cosmo_run, because current post() chains do not contain sampler info, so get_sampler_temperature in getdist fails. I can easily make getdist more robust, but possible more general issue of propagating temperature through post (and awkwardness of having chain metadata stored at sampler level)

**Comment by JesusTorrado** [comment]
> Sounds fine to me, though would be a potentially breaking change having getdist by default. I might be inclined to make a separate method, e.g. mcmc.getdist_chains(...) or similar returning the getdist object directly, avoiding to have to "sample" index the dict. (singular "sample" has always been a bit odd)
> 
> Btw, I don't think we have any function to get mcsamples with the multi-chain information (e.g. for running Getdist's other convergence tests).

Last commit should implement both: separate getdist chains and a `samples` method with a `to_getdist` arg (since we may interface other analysis codes in the future, e.g. arviz).

File: cobaya/prior.py [hunk 1]
```diff
@@ -619,17 +654,16 @@ def set_reference(self, ref_info):
                                   ", a list of two numbers for normal mean and deviation,"
                                   "or a dict with parameters for a scipy distribution.")
         # Re-set the pointlike-ref property
-        if hasattr(self, "_ref_is_pointlike"):
-            delattr(self, "_ref_is_pointlike")
-        self.reference_is_pointlike
+        self._ref_is_pointlike = None
+        self.reference_is_pointlike  # pylint: disable=pointless-statement
```
[the remark this answers is not in the document]
Comment (JesusTorrado) [reply cobaya/prior.py:658]: Changed to yours for easier merge.

File: cobaya/prior.py [hunk 2]
```diff
@@ -547,9 +582,9 @@ def logps_internal(self, x: np.ndarray) -> float:
         """
         self.log.debug("Evaluating prior at %r", x)
         if all(x <= self._upper_limits) and all(x >= self._lower_limits):
-            logps = self._uniform_logp + (sum([logpdf(xi) for logpdf, xi in
-                                               zip(self._non_uniform_logpdf,
-                                                   x[self._non_uniform_indices])])
+            logps = self._uniform_logp + (sum(logpdf(xi) for logpdf, xi in
```
[the remark this answers is not in the document]
Comment (JesusTorrado) [reply cobaya/prior.py:585]: Reverted then. Apparently starts being faster for iterator for very long lists (not the case here).

File: docs/sampler_mcmc.rst [hunk 3]
```diff
@@ -182,6 +182,23 @@ E.g. if a likelihood depends of parameters ``a``, ``b`` and ``c`` and the cost o
 
    If automatic learning of the proposal covariance is enabled, after some checkpoint the proposed steps will mix parameters from different blocks, but *always towards faster ones*. Thus, it is important to specify your blocking in **ascending order of speed**, when not prevented by the architecture of your likelihood (e.g. due to internal caching of intermediate results that require some particular order of parameter variation).
 
+
+.. _mcmc_tempered:
+
+Tempered MCMC
+-------------
+
+Some times it is convenient to sample from a power-reduced (or softened), *tempered* version of the posterior. This produces a Monte Carlo sample with more points towards the tails of the distribution, which is useful when e.g. estimating a quantity by weighting with samples with their probability, or to be able to do more robust :doc:`importance reweighting <post>`.
```
[the remark this answers is not in the document]
Comment (JesusTorrado) [reply docs/sampler_mcmc.rst:191]: Ok.

**Comment by JesusTorrado** [comment]
> Currently the updated getdist does not run with master-branch cobaya, e.g. test_cosmo_run, because current post() chains do not contain sampler info, so get_sampler_temperature in getdist fails. I can easily make getdist more robust, but possible more general issue of propagating temperature through post (and awkwardness of having chain metadata stored at sampler level)

You are right, and the current approach is not correct (my bad), since a sample may have been cooled in Cobaya's side anyway, so the *sampler* temperature and the *samples* temperature may not coincide. Looking into it.

File: cobaya/samplers/mcmc/mcmc.py [hunk 4]
```diff
@@ -857,6 +867,69 @@ def converge_info_changed(self, old_info, new_info):
...
+        combined: bool, default: False
+            If ``True`` and running more than one MPI process, the ``sample`` key of the
+            returned dictionary contains a concatenated sample including all parallel
+            chains concatenated, instead of the chain of the current process only. For
+            this to work, this method needs to be called from all MPI processes
+            simultaneously.
+        skip_samples: int or float, default: 0
+            Skips some amount of initial samples (if ``int``), or an initial fraction of
+            them (if ``float < 1``). If concatenating (``combined=True``), skipping is
+            applied previously to concatenation. Forces the return of a copy.
```
Comment (cmbant) [inline cobaya/samplers/mcmc/mcmc.py:889]: "previously to" -> before

File: tests/common_sampler.py [hunk 5]
```diff
@@ -79,31 +79,30 @@ def body_of_sampler_test(info_sampler: SamplersDict, dimension=1, n_modes=1, tmp
...
-            clusters = [MCSamplesFromCobaya(
-                updated_info, products["clusters"][i]["sample"],
-                name_tag="cluster %d" % (i + 1))
-                for i in products["clusters"]]
+            clusters = [
+                collection["sample"].to_getdist(label="cluster %d" % (i + 1))
+                for i, collection in products["clusters"].items()]
         # Plots!
-        if not is_travis():
+        if do_plots and not is_travis():
```
Comment (cmbant) [inline tests/common_sampler.py:100]: didn't follow why removed?

**Comment by cmbant** [comment]
Getdist updated now, and now seems no getdist-related errors in the re-run log
https://app.travis-ci.com/github/CobayaSampler/cobaya/jobs/597137651
(but several others..)

File: tests/common_sampler.py (hunk 5 above)
Comment (JesusTorrado) [reply tests/common_sampler.py:100]: Makes the text 2x slower when run locally.

**Comment by JesusTorrado** [comment]
> Getdist updated now, and now seems no getdist-related errors in the re-run log https://app.travis-ci.com/github/CobayaSampler/cobaya/jobs/597137651 (but several others..)

It's the MPI post test failing. Checking.

**Pull request merged**